### PR TITLE
Added Check to CeloDistributionSchedule `activate` function

### PR DESCRIPTION
**Pull request by soloseng**
### Description

~~Added a check that prevents activating the `CeloDistributionSchedule` contract if the distributionSchedule contract address if not set in CELO token contract.~~

Previously, it was expected that the `CeloDistributionSchedule` address would be manually set in the `GoldToken` contract by calling `setCeloTokenDistributionSchedule()`. This was expected to be done before calling `activate()` in the `CeloDistributionSchedule`. However, the `CeloDistributionSchedule` contract did not verify that the `CeloDistributionSchedule` address was set in the `GoldToken` contract, allowing for the `CeloDistributionSchedule` to be incorrectly activated.

These new changes get the `CeloDistributionSchedule` address from the registry, instead of setting it in the `GoldToken` contract. Hence, no longer requiring that the `CeloDistributionSchedule` address be manually set in the `GoldToken` contract before activating the `CeloDistributionSchedule` contract.

### Tested

unit tested

### CODE REVIEW

File: packages/protocol/contracts/common/GoldToken.sol [hunk 1]
```diff
@@ -174,8 +153,8 @@ contract GoldToken is
   function transferFrom(address from, address to, uint256 value) external returns (bool) {
     require(to != address(0), "transfer attempted to reserved address 0x0");
     require(
-      to != address(celoTokenDistributionSchedule),
-      "transfer attempted to reserved celoTokenDistributionSchedule address"
+      to != registry.getAddressForOrDie(CELO_DISTRIBUTION_SCHEDULE_ID),
```
Comment (pahor167) [inline packages/protocol/contracts/common/GoldToken.sol:156]: Should we rather use getCeloDistributionSchedule ?

File: packages/protocol/contracts-0.8/common/CeloDistributionSchedule.sol [hunk 2]
```diff
@@ -72,10 +72,14 @@ contract CeloDistributionSchedule is UsingRegistry, ReentrancyGuard, Initializab
     require(address(this).balance > 0, "Contract does not have CELO balance.");
     require(!areDependenciesSet, "Contract has already been activated.");
     require(block.timestamp > _l2StartTime, "L2 start time cannot be set to a future date.");
+    ICeloToken celoToken = ICeloToken(address(getCeloToken()));
+    require(
+      registry.getAddressForOrDie(CELO_DISTRIBUTION_SCHEDULE_ID) == address(this),
+      "CeloDistributionSchedule address is incorrectly set in Registry."
+    );
```
Comment (arthurgousset) [inline packages/protocol/contracts-0.8/common/CeloDistributionSchedule.sol:79]: +1 to @pahor167's comment. Might be more readable to use `getCeloDistributionSchedule()` from `UsingRegistry.sol`:

```suggestion
    require(
      address(getCeloDistributionSchedule()) == address(this),
      "CeloDistributionSchedule address is incorrectly set in Registry."
    );
```

File: packages/protocol/contracts/common/GoldToken.sol [hunk 3]
```diff
@@ -269,6 +248,13 @@ contract GoldToken is
     return allowed[_owner][spender];
   }
 
+  /**
+   * @return The address of the CeloTokenDistributionSchedule contract.
+   */
+  function getCeloTokenDistributionScheduleAddress() external view returns (address) {
+    return registry.getAddressForOrDie(CELO_DISTRIBUTION_SCHEDULE_ID);
+  }
```
Comment (arthurgousset) [inline packages/protocol/contracts/common/GoldToken.sol]: Nit: Unless I'm missing something, this function doesn't seem to be used anywhere in this contract or elsewhere. 
I propose we remove it and use something like `address(getCeloDistributionSchedule())` anywhere it might be needed.

If we remove the function here, we should also remove it from the [`ICeloToken.sol` interface](https://github.com/celo-org/celo-monorepo/blob/015002a5deff86d5ab646ee811fdabf133fbd4d0/packages/protocol/contracts-0.8/common/interfaces/ICeloToken.sol#L26).

File: packages/protocol/test-sol/unit/common/CeloDistributionSchedule.t.sol [hunk 4]
```diff
@@ -302,6 +296,50 @@ contract CeloDistributionScheduleTest_activate is CeloDistributionScheduleTest {
...
+      carbonOffsettingPartner,
+      carbonOffsettingFraction
+    );
+  }
+
+  function test_Reverts_WhenCeloDistributionAddressIncorrectlySetInRegistry() public {
+    vm.warp(block.timestamp + l2StartTime);
+    vm.prank(celoDistributionOwner);
+    celoDistributionSchedule = new CeloDistributionSchedule(true);
+    registry.setAddressFor("CeloDistributionSchedule", celoTokenAddress);
```
Comment (arthurgousset) [inline packages/protocol/test-sol/unit/common/CeloDistributionSchedule.t.sol]: Nit: While reading the test, I'd find it easier to have an arbitrary wrong address, instead of the `CeloToken` address. 
For context, I caught myself wondering whether it's actually correct that this is the `CeloToken` address, or whether it was intentionally wrong. With a little more thinking, it's clear that this is intentionally wrong, but it wasn't obvious from reading `celoTokenAddress`.

```suggestion
    registry.setAddressFor("CeloDistributionSchedule", actor("arbitraryWrongAddress"));
```

File: packages/protocol/test-sol/unit/common/FeeHandler.t.sol [hunk 5]
```diff
@@ -102,6 +103,7 @@ contract FeeHandlerTest is Test, Constants {
     registry.setAddressFor("GoldToken", address(celoToken));
     registry.setAddressFor("CeloToken", address(celoToken));
     registry.setAddressFor("Reserve", address(mockReserve));
+    registry.setAddressFor("CeloDistributionSchedule", address(celoDistributionSchedule));
```
Comment (arthurgousset) [inline packages/protocol/test-sol/unit/common/FeeHandler.t.sol]: Nit: `celoDistributionSchedule` is already an `address` from here `address celoDistributionSchedule = actor("CeloDistributionSchedule")`.

```suggestion
    registry.setAddressFor("CeloDistributionSchedule", celoDistributionSchedule);
```

File: packages/protocol/test-sol/unit/governance/voting/ReleaseGold.t.sol [hunk 6]
```diff
@@ -136,6 +124,7 @@ contract ReleaseGoldTest is Test, ECDSAHelper {
     registry.setAddressFor("LockedGold", address(lockedGold));
     registry.setAddressFor("Validators", address(validators));
     registry.setAddressFor("StableToken", address(stableToken));
+    registry.setAddressFor("CeloDistributionSchedule", address(celoDistributionSchedule));
```
Comment (arthurgousset) [inline packages/protocol/test-sol/unit/governance/voting/ReleaseGold.t.sol]: Nit: `celoDistributionSchedule` is already an address from `address celoDistributionSchedule = actor("CeloDistributionSchedule")`

```suggestion
    registry.setAddressFor("CeloDistributionSchedule", celoDistributionSchedule);
```

**Review by arthurgousset — DISMISSED** [review]
Left a couple nit comments, but looks good to me. I like how you wrote the two new unit tests. It's clever to expect the contract to revert.

Nit: `registryAddress` and `proxyAdminAddress` (as examples) are constants that could live in [`test-sol/constants.sol`](https://github.com/celo-org/celo-monorepo/blob/e7710e29bb000ad575914ba3884cc92bda58b768/packages/protocol/test-sol/constants.sol). Probably not worth refactoring in this PR, but something that might be worth doing separately.

**Review by arthurgousset — COMMENTED** [review]
I'm not 100% sure I understand the original bug, and how this PR fixes it, so not super comfortable approving. But left some nit comments from reading through the PR.

File: packages/protocol/contracts/common/GoldToken.sol (hunk 1 above)
Comment (soloseng) [reply packages/protocol/contracts/common/GoldToken.sol:156]: I ended up using `registry.getAddressForOrDie(CELO_DISTRIBUTION_SCHEDULE_ID)` to optimize for gas cost.

Otherwise, we convert an address to an interface and back to an address.

Comment (soloseng) [reply packages/protocol/contracts/common/GoldToken.sol:156]: I can change it if you think its necessary.

File: packages/protocol/contracts-0.8/common/CeloDistributionSchedule.sol (hunk 2 above)
Comment (soloseng) [reply packages/protocol/contracts-0.8/common/CeloDistributionSchedule.sol:79]: see [comment](https://github.com/celo-org/celo-monorepo/pull/11109#discussion_r1678067925) above

**Comment by soloseng** [comment]
> I'm not 100% sure I understand the original bug, and how this PR fixes it, so not super comfortable approving. But left some nit comments from reading through the PR.

Updated the PR description to make it more clear what the changes are and why they were needed.

File: packages/protocol/contracts-0.8/common/CeloDistributionSchedule.sol
Comment (arthurgousset) [inline packages/protocol/contracts-0.8/common/CeloDistributionSchedule.sol:1]: There is a [CI error](https://github.com/celo-org/celo-monorepo/actions/runs/9978749616/job/27576665478?pr=11109), because it cannot find the `celoDistributionSchedule` artifacts. Using this thread to keep track of findings.

Comment (arthurgousset) [reply packages/protocol/contracts-0.8/common/CeloDistributionSchedule.sol:1]: I'm able to reproduce this locally: 

```sh
$ yarn --cwd packages/protocol test common/

# ...
Error: Could not find artifacts for CeloDistributionSchedule from any sources
    at Resolver.require (webpack:///./packages/resolver/dist/lib/resolver.js:53:1)
    at Object.require (webpack:///./packages/test/dist/src/Test.js:276:1)
    at Object.<anonymous> (/Users/arthur/Documents/celo-org/celo-monorepo/packages/protocol/test/common/feehandlerseller.ts:27:78)
    at Module._compile (node:internal/modules/cjs/loader:1364:14)
    at Module.m._compile (/Users/arthur/Documents/celo-org/celo-monorepo/node_modules/ts-node/src/index.ts:1618:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1422:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/arthur/Documents/celo-org/celo-monorepo/node_modules/ts-node/src/index.ts:1621:12)
    at Module.load (node:internal/modules/cjs/loader:1203:32)
    at Function.Module._load (node:internal/modules/cjs/loader:1019:12)
    at Module.require (node:internal/modules/cjs/loader:1231:19)
    at require (node:internal/modules/helpers:177:18)
    at /Users/arthur/Documents/celo-org/celo-monorepo/node_modules/truffle/node_modules/mocha/lib/mocha.js:414:36
    at Array.forEach (<anonymous>)
    at Mocha.loadFiles (/Users/arthur/Documents/celo-org/celo-monorepo/node_modules/truffle/node_modules/mocha/lib/mocha.js:411:14)
    at Mocha.run (/Users/arthur/Documents/celo-org/celo-monorepo/node_modules/truffle/node_modules/mocha/lib/mocha.js:969:10)
    at webpack:///./packages/test/dist/src/Test.js:173:1
    at new Promise (<anonymous>)
    at Object.<anonymous> (webpack:///./packages/test/dist/src/Test.js:172:1)
    at Generator.next (<anonymous>)
    at fulfilled (webpack:///./packages/test/dist/src/Test.js:28:43)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
Truffle v5.9.0 (core: 5.9.0)
Node v18.20.2
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
{
  code: 1,
  stout: 'bigint: Failed to load bindings, pure JS will be used (try npm run rebuild?)\n' +
    ' \x1B[2K \x1B[1G\x1B[31merror\x1B[39m Command failed with exit code 1.\n'
}
error Command failed with exit code 1.
```

Comment (arthurgousset) [reply packages/protocol/contracts-0.8/common/CeloDistributionSchedule.sol:1]: The error might come from here: 

https://github.com/celo-org/celo-monorepo/blob/01174be3e4d946232ad718c36352ba7382132ba7/packages/protocol/test/common/feehandlerseller.ts#L17

Yet, during the migrations the artifact is found correctly here:

https://github.com/celo-org/celo-monorepo/blob/01174be3e4d946232ad718c36352ba7382132ba7/packages/protocol/migrations_ts/28_celoDistributionSchedule.ts#L7

```sh
-> Deploying CeloDistributionSchedule
  Setting initial CeloDistributionSchedule implementation on proxy
```

Comment (arthurgousset) [reply packages/protocol/contracts-0.8/common/CeloDistributionSchedule.sol:1]: I tried to poke around and see what might cause this. Couldn't find anything that jumped out at me. I'll pause here.

Comment (soloseng) [reply packages/protocol/contracts-0.8/common/CeloDistributionSchedule.sol:1]: The issue was because the solidity contracts with `pragma < 0.8` and use a different artifact folder when they are built by truffle. I updated it in https://github.com/celo-org/celo-monorepo/pull/11109/commits/e6cbe0878df54496f304fea688a35488ba280df5 and it works now.

File: packages/protocol/contracts/common/GoldToken.sol (hunk 1 above)
Comment (arthurgousset) [reply packages/protocol/contracts/common/GoldToken.sol:156]: For context, what's the difference in gas cost (if that's easy to calculate)? I have no intuition for that. 

Personally, I'd prioritise contract readability over gas efficiency, but that's coming from a place of "maintaining" the contracts, and less of "using" the contracts. So I'll defer judgement to @pahor167 and you on this @soloseng.

**Review by arthurgousset — APPROVED** [review]
Pending this [conversation](https://github.com/celo-org/celo-monorepo/pull/11109#discussion_r1683162811) the changes look good to me. But, I'm always a little nervous approving changes to core contracts. So take my approval with a pinch of salt. I'd prefer if @martinvol or @pahor167 gave this an approval too.

File: packages/protocol/contracts/common/GoldToken.sol (hunk 1 above)
Comment (martinvol) [reply packages/protocol/contracts/common/GoldToken.sol:156]: I'm not sure if it saves gas but I like it because it helps us move in this direction https://github.com/celo-org/celo-monorepo/issues/11151

File: packages/protocol/test/common/feehandlerseller.ts [hunk 7]
```diff
@@ -23,24 +26,32 @@ const UniswapFeeHandlerSeller: UniswapFeeHandlerSellerContract =
...
   beforeEach(async () => {
     registry = await Registry.new(true)
+    celoDistributionSchedule = await CeloDistributionSchedule.new(true)
 
     goldToken = await GoldToken.new(true)
     await goldToken.initialize(registry.address)
     await registry.setAddressFor(CeloContractName.GoldToken, goldToken.address)
+    await registry.setAddressFor(
+      CeloContractName.CeloDistributionSchedule,
+      celoDistributionSchedule.address
```
Comment (martinvol) [inline packages/protocol/test/common/feehandlerseller.ts:53]: why is this needed? A comment would help

**Review by martinvol — APPROVED** [review]

File: packages/protocol/test/common/feehandlerseller.ts (hunk 7 above)
Comment (soloseng) [reply packages/protocol/test/common/feehandlerseller.ts:53]: Added because the CeloToken `_transfer` prevents transfers to the celoDistributionSchedule, so it needs to be set in the precompile otherwise transfers revert.

File: packages/protocol/contracts/common/GoldToken.sol (hunk 1 above)
Comment (soloseng) [reply packages/protocol/contracts/common/GoldToken.sol:156]: > For context, what's the difference in gas cost (if that's easy to calculate)? I have no intuition for that.

I usually check by running the test in foundry using both methods when I want to have an idea of which would be higher.
